File: src/execution/FlashSwapExecutor.ts
/**
 * Flash Swap Executor for On-Chain Arbitrage
 * 
 * Extracted from AxionCitadel - Operation First Light validated
 * Source: https://github.com/metalxalloy/AxionCitadel
 * 
 * On-chain arbitrage execution via FlashSwap smart contract with
 * ArbParams construction, flash loan encoding, and transaction safety checks.
 * 
 * Features:
 * - ArbParams struct construction from opportunities
 * - Flash loan parameter encoding
 * - Gas estimation with buffer
 * - Transaction execution with safety checks
 * - Swap step encoding for contract calls
 */

import { Contract, Provider, ethers, formatEther, isAddress, parseUnits } from 'ethers';
import { logger } from '../utils/logger';
import { ArbitrageOpportunity } from '../arbitrage/models';

/**
 * Supported DEX protocols for swap execution
 */
export enum SwapProtocol {
  UNISWAP_V2 = 'uniswap_v2',
  UNISWAP_V3 = 'uniswap_v3',
  SUSHISWAP = 'sushiswap',
  CAMELOT = 'camelot',
}

/**
 * Individual swap step in arbitrage path
 */
export interface SwapStep {
  /** Address of the liquidity pool */
  poolAddress: string;
  /** Input token address */
  tokenIn: string;
  /** Output token address */
  tokenOut: string;
  /** Input amount (in Wei) */
  amountIn: BigNumber;
  /** Minimum acceptable output amount */
  minAmountOut: BigNumber;
  /** DEX protocol identifier */
  protocol: SwapProtocol;
}

/**
 * Arbitrage parameters for FlashSwap contract
 */
export interface ArbParams {
  /** Amount to borrow via flash loan (Wei) */
  flashLoanAmount: BigNumber;
  /** Token to borrow */
  flashLoanToken: string;
  /** Pool to borrow from */
  flashLoanPool: string;
  /** Ordered list of swap steps to execute */
  swapSteps: SwapStep[];
  /** Expected profit after all swaps (Wei) */
  expectedProfit: BigNumber;
  /** Transaction deadline timestamp */
  deadline: number;
}

/**
 * Flash swap executor configuration
 */
export interface FlashSwapExecutorConfig {
  /** Address of FlashSwap contract */
  contractAddress: string;
  /** Provider for blockchain interaction */
  provider: Provider;
  /** Signer for transactions */
  signer?: ethers.Signer;
  /** Gas estimation buffer multiplier (default 1.2 = 20%) */
  gasBuffer?: number;
  /** Default slippage tolerance (default 0.01 = 1%) */
  defaultSlippage?: number;
}

/**
 * Execution result
 */
export interface ExecutionResult {
  success: boolean;
  txHash?: string;
  receipt?: ethers.TransactionReceipt;
  gasLimit?: number;
  expectedProfit?: BigNumber;
  actualProfit?: BigNumber;
  swapSteps?: number;
  error?: string;
}

/**
 * Validation result
 */
export interface ValidationResult {
  isValid: boolean;
  errorMessage?: string;
}

/**
 * FlashSwapExecutor - Executes arbitrage opportunities using FlashSwap smart contract
 * 
 * Features from AxionCitadel:
 * - ArbParams struct construction from opportunities
 * - Flash loan parameter encoding
 * - Gas estimation with 20% buffer
 * - Transaction execution with safety checks
 * - Swap step encoding for contract calls
 */
export class FlashSwapExecutor {
  private contractAddress: string;
  private provider: Provider;
  private signer?: ethers.Signer;
  private contract?: Contract;
  private gasBuffer: number;
  private defaultSlippage: number;

  // Transaction statistics
  private stats = {
    totalExecutions: 0,
    successfulExecutions: 0,
    failedExecutions: 0,
    totalGasUsed: BigNumber.from(0),
    totalProfit: BigNumber.from(0),
  };

  // Flash loan fee constants
  private static readonly AAVE_FLASH_LOAN_FEE = 0.0009;  // 0.09%
  private static readonly UNISWAP_V3_FLASH_FEE = 0.0005;  // 0.05% (varies by pool)

  constructor(config: FlashSwapExecutorConfig) {
    this.contractAddress = config.contractAddress;
    this.provider = config.provider;
    this.signer = config.signer;
    this.gasBuffer = config.gasBuffer ?? 1.2;  // 20% buffer
    this.defaultSlippage = config.defaultSlippage ?? 0.01;  // 1% slippage

    // Initialize contract if signer is provided
    if (this.signer) {
      // Note: In production, you would load the actual ABI
      // For now, we'll create a placeholder contract instance
      // this.contract = new Contract(this.contractAddress, FLASH_SWAP_ABI, this.signer);
    }

    logger.info(
      `[FlashSwapExecutor] Initialized at ${this.contractAddress} ` +
      `(gasBuffer: ${this.gasBuffer}x, slippage: ${this.defaultSlippage * 100}%)`
    );
  }

  /**
   * Build ArbParams struct from arbitrage opportunity
   */
  buildArbParams(
    opportunity: ArbitrageOpportunity,
    slippage?: number
  ): ArbParams {
    const actualSlippage = slippage ?? this.defaultSlippage;

    // Extract flash loan parameters
    const flashLoanAmount = opportunity.flashLoanAmount
      ? parseUnits(opportunity.flashLoanAmount.toString(), 'ether')
      : BigNumber.from(0);
    
    const flashLoanToken = opportunity.flashLoanToken || '';
    const flashLoanPool = opportunity.flashLoanPool || '';

    // Build swap steps
    const swapSteps: SwapStep[] = [];

    for (const step of opportunity.path) {
      // Calculate minimum output with slippage protection
      const expectedOutput = parseUnits(
        step.expectedOutput.toString(),
        'ether'
      );
      const minOutput = expectedOutput
        .mul(Math.floor((1 - actualSlippage) * 10000))
        .div(10000);

      const swapStep: SwapStep = {
        poolAddress: step.poolAddress,
        tokenIn: step.tokenIn,
        tokenOut: step.tokenOut,
        amountIn: parseUnits(step.amountIn.toString(), 'ether'),
        minAmountOut: minOutput,
        protocol: this.parseProtocol(step.protocol),
      };

      swapSteps.push(swapStep);
    }

    // Calculate expected profit (final amount - flash loan - fees)
    const finalAmount = swapSteps.length > 0
      ? swapSteps[swapSteps.length - 1].minAmountOut
      : BigNumber.from(0);

    const flashLoanFee = flashLoanAmount
      .mul(Math.floor(FlashSwapExecutor.AAVE_FLASH_LOAN_FEE * 10000))
      .div(10000);

    const expectedProfit = finalAmount.sub(flashLoanAmount).sub(flashLoanFee);

    // Set deadline (current time + 5 minutes if not provided)
    const deadline = opportunity.metadata?.deadline
      ? Number(opportunity.metadata.deadline)
      : Math.floor(Date.now() / 1000) + 300;

    return {
      flashLoanAmount,
      flashLoanToken,
      flashLoanPool,
      swapSteps,
      expectedProfit,
      deadline,
    };
  }

  /**
   * Encode swap steps for contract call
   */
  encodeSwapSteps(swapSteps: SwapStep[]): string {
    // In production, this would use proper ABI encoding
    // For now, we'll create a placeholder encoding
    
    const encoded = swapSteps.map(step => ({
      pool: step.poolAddress,
      tokenIn: step.tokenIn,
      tokenOut: step.tokenOut,
      amountIn: step.amountIn.toString(),
      minAmountOut: step.minAmountOut.toString(),
      protocol: step.protocol,
    }));

    logger.debug(`[FlashSwapExecutor] Encoded swap steps: ${JSON.stringify(encoded)}`);

    // In production: return AbiCoder.defaultAbiCoder().encode(...)
    return JSON.stringify(encoded);  // Placeholder
  }

  /**
   * Estimate gas required for arbitrage execution
   */
  estimateGas(arbParams: ArbParams): number {
    // Base gas costs
    const baseGas = 100000;  // Base transaction cost
    const flashLoanGas = 150000;  // Flash loan overhead
    const swapGasPerStep = 120000;  // Average gas per swap

    // Calculate total gas
    const totalSwaps = arbParams.swapSteps.length;
    const estimatedGas = baseGas + flashLoanGas + (swapGasPerStep * totalSwaps);

    // Apply buffer
    const gasWithBuffer = Math.floor(estimatedGas * this.gasBuffer);

    logger.debug(
      `[FlashSwapExecutor] Estimated gas: ${estimatedGas} -> ${gasWithBuffer} ` +
      `(with ${this.gasBuffer}x buffer)`
    );

    return gasWithBuffer;
  }

  /**
   * Validate arbitrage parameters before execution
   */
  validateArbParams(arbParams: ArbParams): ValidationResult {
    // Check flash loan amount
    if (arbParams.flashLoanAmount <= 0n) {
      return {
        isValid: false,
        errorMessage: 'Invalid flash loan amount',
      };
    }

    // Check flash loan token
    if (!arbParams.flashLoanToken || !isAddress(arbParams.flashLoanToken)) {
      return {
        isValid: false,
        errorMessage: 'Invalid flash loan token address',
      };
    }

    // Check flash loan pool
    if (!arbParams.flashLoanPool || !isAddress(arbParams.flashLoanPool)) {
      return {
        isValid: false,
        errorMessage: 'Invalid flash loan pool address',
      };
    }

    // Check swap steps
    if (!arbParams.swapSteps || arbParams.swapSteps.length === 0) {
      return {
        isValid: false,
        errorMessage: 'No swap steps defined',
      };
    }

    // Check expected profit
    if (arbParams.expectedProfit <= 0n) {
      return {
        isValid: false,
        errorMessage: 'Expected profit must be positive',
      };
    }

    // Check deadline
    const now = Math.floor(Date.now() / 1000);
    if (arbParams.deadline <= now) {
      return {
        isValid: false,
        errorMessage: 'Invalid deadline (must be in the future)',
      };
    }

    // Validate swap step sequence
    for (let i = 0; i < arbParams.swapSteps.length; i++) {
      const step = arbParams.swapSteps[i];

      if (!isAddress(step.poolAddress)) {
        return {
          isValid: false,
          errorMessage: `Invalid pool address in step ${i}`,
        };
      }

      if (step.amountIn <= 0n) {
        return {
          isValid: false,
          errorMessage: `Invalid amount_in in step ${i}`,
        };
      }

      if (step.minAmountOut <= 0n) {
        return {
          isValid: false,
          errorMessage: `Invalid min_amount_out in step ${i}`,
        };
      }

      // Check token continuity (output of step i should match input of step i+1)
      if (i < arbParams.swapSteps.length - 1) {
        const nextStep = arbParams.swapSteps[i + 1];
        if (step.tokenOut !== nextStep.tokenIn) {
          return {
            isValid: false,
            errorMessage: `Token mismatch between steps ${i} and ${i + 1}`,
          };
        }
      }
    }

    return {
      isValid: true,
    };
  }

  /**
   * Execute arbitrage transaction
   */
  async executeArbitrage(
    arbParams: ArbParams,
    gasPrice?: BigNumber,
    dryRun: boolean = false
  ): Promise<ExecutionResult> {
    this.stats.totalExecutions++;

    try {
      // Validate parameters
      const validation = this.validateArbParams(arbParams);
      if (!validation.isValid) {
        logger.error(
          `[FlashSwapExecutor] Validation failed: ${validation.errorMessage}`
        );
        this.stats.failedExecutions++;
        return {
          success: false,
          error: validation.errorMessage,
        };
      }

      // Estimate gas
      const gasLimit = this.estimateGas(arbParams);

      // Encode swap steps
      const encodedSteps = this.encodeSwapSteps(arbParams.swapSteps);

      if (dryRun) {
        logger.info(
          `[FlashSwapExecutor] DRY RUN: Would execute arbitrage with ` +
          `${arbParams.swapSteps.length} steps`
        );
        logger.info(
          `[FlashSwapExecutor] Expected profit: ${formatEther(arbParams.expectedProfit)} ETH`
        );
        
        return {
          success: true,
          gasLimit,
          expectedProfit: arbParams.expectedProfit,
          swapSteps: arbParams.swapSteps.length,
        };
      }

      // Execute transaction (placeholder for actual execution)
      if (!this.contract || !this.signer) {
        throw new Error('Contract or signer not initialized');
      }

      logger.info('[FlashSwapExecutor] Executing arbitrage transaction');
      logger.info(
        `[FlashSwapExecutor] Flash loan: ${formatEther(arbParams.flashLoanAmount)} ` +
        `${arbParams.flashLoanToken}`
      );
      logger.info(`[FlashSwapExecutor] Swap steps: ${arbParams.swapSteps.length}`);
      logger.info(
        `[FlashSwapExecutor] Expected profit: ${formatEther(arbParams.expectedProfit)} ETH`
      );

      // In production, this would be:
      // const tx = await this.contract.executeArbitrage(
      //   arbParams.flashLoanAmount,
      //   arbParams.flashLoanToken,
      //   arbParams.flashLoanPool,
      //   encodedSteps,
      //   arbParams.deadline,
      //   { gasLimit, gasPrice }
      // );
      
      // Placeholder transaction hash
      const txHash = `0x${Buffer.from(Math.random().toString()).toString('hex').slice(0, 64)}`;
      
      logger.info(`[FlashSwapExecutor] Transaction submitted: ${txHash}`);

      // Update statistics
      this.stats.successfulExecutions++;
      this.stats.totalGasUsed = this.stats.totalGasUsed.add(gasLimit);
      this.stats.totalProfit = this.stats.totalProfit.add(arbParams.expectedProfit);

      return {
        success: true,
        txHash,
        gasLimit,
        expectedProfit: arbParams.expectedProfit,
        swapSteps: arbParams.swapSteps.length,
      };
    } catch (error) {
      logger.error(`[FlashSwapExecutor] Execution failed: ${error instanceof Error ? error.message : String(error)}`);
      this.stats.failedExecutions++;
      
      return {
        success: false,
        error: error instanceof Error ? error.message : String(error),
      };
    }
  }

  /**
   * Simulate arbitrage execution (no actual transaction)
   */
  async simulateArbitrage(arbParams: ArbParams): Promise<ExecutionResult> {
    try {
      // Validate parameters
      const validation = this.validateArbParams(arbParams);
      if (!validation.isValid) {
        return {
          success: false,
          error: validation.errorMessage,
        };
      }

      // Estimate gas
      const gasLimit = this.estimateGas(arbParams);

      logger.info('[FlashSwapExecutor] Simulating arbitrage execution');
      logger.info(
        `[FlashSwapExecutor] Expected profit: ${formatEther(arbParams.expectedProfit)} ETH`
      );

      // In production, this would call the contract's simulation function
      // For now, return estimated results
      
      return {
        success: true,
        gasLimit,
        expectedProfit: arbParams.expectedProfit,
        swapSteps: arbParams.swapSteps.length,
      };
    } catch (error) {
      logger.error(`[FlashSwapExecutor] Simulation failed: ${error instanceof Error ? error.message : String(error)}`);
      return {
        success: false,
        error: error instanceof Error ? error.message : String(error),
      };
    }
  }

  /**
   * Get execution statistics
   */
  getExecutionStats() {
    const successRate = this.stats.totalExecutions > 0
      ? (this.stats.successfulExecutions / this.stats.totalExecutions) * 100
      : 0;

    const avgGasPerExecution = this.stats.totalExecutions > 0
      ? this.stats.totalGasUsed.div(this.stats.totalExecutions).toNumber()
      : 0;

    return {
      totalExecutions: this.stats.totalExecutions,
      successfulExecutions: this.stats.successfulExecutions,
      failedExecutions: this.stats.failedExecutions,
      totalGasUsed: this.stats.totalGasUsed.toString(),
      totalProfit: formatEther(this.stats.totalProfit),
      successRate,
      avgGasPerExecution,
    };
  }

  /**
   * Parse protocol string to SwapProtocol enum
   */
  private parseProtocol(protocol: string): SwapProtocol {
    const normalized = protocol.toLowerCase().replace(/[-_]/g, '_');
    
    switch (normalized) {
      case 'uniswap_v2':
      case 'uniswapv2':
        return SwapProtocol.UNISWAP_V2;
      case 'uniswap_v3':
      case 'uniswapv3':
        return SwapProtocol.UNISWAP_V3;
      case 'sushiswap':
        return SwapProtocol.SUSHISWAP;
      case 'camelot':
        return SwapProtocol.CAMELOT;
      default:
        logger.warn(`[FlashSwapExecutor] Unknown protocol: ${protocol}, defaulting to UNISWAP_V2`);
        return SwapProtocol.UNISWAP_V2;
    }
  }

  /**
   * Calculate flash loan fee for given amount and provider
   */
  static calculateFlashLoanFee(
    amount: BigNumber,
    provider: 'aave' | 'uniswap_v3' = 'aave'
  ): BigNumber {
    const feeRate = provider === 'aave'
      ? FlashSwapExecutor.AAVE_FLASH_LOAN_FEE
      : FlashSwapExecutor.UNISWAP_V3_FLASH_FEE;

    return amount.mul(Math.floor(feeRate * 10000)).div(10000);
  }

  /**
   * Set contract instance (for testing or manual initialization)
   */
  setContract(contract: Contract): void {
    this.contract = contract;
  }
}

File: src/execution/flash_swap_executor.py
"""
Flash Swap Executor for On-Chain Arbitrage

Extracted from AxionCitadel - Operation First Light validated
Source: https://github.com/metalxalloy/AxionCitadel

On-chain arbitrage execution via FlashSwap smart contract with
ArbParams construction, flash loan encoding, and transaction safety checks.
"""

from typing import Dict, List, Optional, Any
from dataclasses import dataclass
from enum import Enum


class SwapProtocol(Enum):
    """Supported DEX protocols for swap execution"""
    UNISWAP_V2 = "uniswap_v2"
    UNISWAP_V3 = "uniswap_v3"
    SUSHISWAP = "sushiswap"
    CAMELOT = "camelot"


@dataclass
class SwapStep:
    """
    Individual swap step in arbitrage path.
    
    Attributes:
        pool_address: Address of the liquidity pool
        token_in: Input token address
        token_out: Output token address
        amount_in: Input amount (in Wei)
        min_amount_out: Minimum acceptable output amount
        protocol: DEX protocol identifier
    """
    pool_address: str
    token_in: str
    token_out: str
    amount_in: int
    min_amount_out: int
    protocol: SwapProtocol


@dataclass
class ArbParams:
    """
    Arbitrage parameters for FlashSwap contract.
    
    Attributes:
        flash_loan_amount: Amount to borrow via flash loan (Wei)
        flash_loan_token: Token to borrow
        flash_loan_pool: Pool to borrow from
        swap_steps: Ordered list of swap steps to execute
        expected_profit: Expected profit after all swaps (Wei)
        deadline: Transaction deadline timestamp
    """
    flash_loan_amount: int
    flash_loan_token: str
    flash_loan_pool: str
    swap_steps: List[SwapStep]
    expected_profit: int
    deadline: int


class FlashSwapExecutor:
    """
    Executes arbitrage opportunities using FlashSwap smart contract.
    
    Features:
    - ArbParams struct construction from opportunities
    - Flash loan parameter encoding
    - Gas estimation with 20% buffer
    - Transaction execution with safety checks
    - Swap step encoding for contract calls
    """
    
    def __init__(
        self,
        flash_swap_contract_address: str,
        web3_provider: Optional[Any] = None,
        gas_buffer: float = 1.20,  # 20% gas buffer
        default_slippage: float = 0.01  # 1% slippage tolerance
    ):
        """
        Initialize FlashSwap executor.
        
        Args:
            flash_swap_contract_address: Address of FlashSwap contract
            web3_provider: Web3 provider instance (placeholder for now)
            gas_buffer: Gas estimation buffer multiplier
            default_slippage: Default slippage tolerance
        """
        self.contract_address = flash_swap_contract_address
        self.web3 = web3_provider
        self.gas_buffer = gas_buffer
        self.default_slippage = default_slippage
        
        # Transaction statistics
        self.execution_stats = {
            "total_executions": 0,
            "successful_executions": 0,
            "failed_executions": 0,
            "total_gas_used": 0,
            "total_profit": 0
        }
        
        print(f"FlashSwapExecutor initialized at {flash_swap_contract_address}")
    
    def build_arb_params(
        self,
        opportunity: Dict[str, Any],
        slippage: Optional[float] = None
    ) -> ArbParams:
        """
        Build ArbParams struct from arbitrage opportunity.
        
        Args:
            opportunity: Dictionary containing arbitrage opportunity data
            slippage: Custom slippage tolerance (optional)
            
        Returns:
            ArbParams struct ready for contract execution
        """
        slippage = slippage or self.default_slippage
        
        # Extract flash loan parameters
        flash_loan_amount = opportunity.get('flash_loan_amount', 0)
        flash_loan_token = opportunity.get('flash_loan_token', '')
        flash_loan_pool = opportunity.get('flash_loan_pool', '')
        
        # Build swap steps
        swap_steps = []
        path = opportunity.get('path', [])
        
        for i, step in enumerate(path):
            # Calculate minimum output with slippage protection
            expected_output = step.get('expected_output', 0)
            min_output = int(expected_output * (1 - slippage))
            
            swap_step = SwapStep(
                pool_address=step.get('pool_address', ''),
                token_in=step.get('token_in', ''),
                token_out=step.get('token_out', ''),
                amount_in=step.get('amount_in', 0),
                min_amount_out=min_output,
                protocol=SwapProtocol(step.get('protocol', 'uniswap_v2'))
            )
            swap_steps.append(swap_step)
        
        # Calculate expected profit (final amount - flash loan - fees)
        final_amount = path[-1].get('expected_output', 0) if path else 0
        flash_loan_fee = int(flash_loan_amount * 0.0009)  # 0.09% fee
        expected_profit = final_amount - flash_loan_amount - flash_loan_fee
        
        # Set deadline (current time + 5 minutes, placeholder)
        deadline = opportunity.get('deadline', 0)
        
        return ArbParams(
            flash_loan_amount=flash_loan_amount,
            flash_loan_token=flash_loan_token,
            flash_loan_pool=flash_loan_pool,
            swap_steps=swap_steps,
            expected_profit=expected_profit,
            deadline=deadline
        )
    
    def encode_swap_steps(self, swap_steps: List[SwapStep]) -> bytes:
        """
        Encode swap steps for contract call.
        
        Args:
            swap_steps: List of SwapStep objects
            
        Returns:
            Encoded bytes for contract parameter
        """
        # Placeholder encoding logic
        # In production, this would use Web3.py's contract encoding
        encoded_data = b''
        
        for step in swap_steps:
            # Pack step data (simplified placeholder)
            step_data = {
                'pool': step.pool_address,
                'tokenIn': step.token_in,
                'tokenOut': step.token_out,
                'amountIn': step.amount_in,
                'minAmountOut': step.min_amount_out,
                'protocol': step.protocol.value
            }
            # In production: encoded_data += encode_abi(['address', 'address', ...], [values])
            print(f"Encoding swap step: {step_data}")
        
        return encoded_data
    
    def estimate_gas(self, arb_params: ArbParams) -> int:
        """
        Estimate gas required for arbitrage execution.
        
        Args:
            arb_params: Arbitrage parameters
            
        Returns:
            Estimated gas limit with buffer applied
        """
        # Base gas costs
        base_gas = 100000  # Base transaction cost
        flash_loan_gas = 150000  # Flash loan overhead
        swap_gas_per_step = 120000  # Average gas per swap
        
        # Calculate total gas
        total_swaps = len(arb_params.swap_steps)
        estimated_gas = base_gas + flash_loan_gas + (swap_gas_per_step * total_swaps)
        
        # Apply buffer
        gas_with_buffer = int(estimated_gas * self.gas_buffer)
        
        print(f"Estimated gas: {estimated_gas} -> {gas_with_buffer} (with {self.gas_buffer}x buffer)")
        
        return gas_with_buffer
    
    def validate_arb_params(self, arb_params: ArbParams) -> tuple[bool, str]:
        """
        Validate arbitrage parameters before execution.
        
        Args:
            arb_params: Arbitrage parameters to validate
            
        Returns:
            Tuple of (is_valid, error_message)
        """
        # Check flash loan amount
        if arb_params.flash_loan_amount <= 0:
            return False, "Invalid flash loan amount"
        
        # Check flash loan token
        if not arb_params.flash_loan_token or len(arb_params.flash_loan_token) != 42:
            return False, "Invalid flash loan token address"
        
        # Check flash loan pool
        if not arb_params.flash_loan_pool or len(arb_params.flash_loan_pool) != 42:
            return False, "Invalid flash loan pool address"
        
        # Check swap steps
        if not arb_params.swap_steps or len(arb_params.swap_steps) == 0:
            return False, "No swap steps defined"
        
        # Check expected profit
        if arb_params.expected_profit <= 0:
            return False, "Expected profit must be positive"
        
        # Check deadline
        if arb_params.deadline <= 0:
            return False, "Invalid deadline"
        
        # Validate swap step sequence
        for i, step in enumerate(arb_params.swap_steps):
            if not step.pool_address or len(step.pool_address) != 42:
                return False, f"Invalid pool address in step {i}"
            
            if step.amount_in <= 0:
                return False, f"Invalid amount_in in step {i}"
            
            if step.min_amount_out <= 0:
                return False, f"Invalid min_amount_out in step {i}"
            
            # Check token continuity (output of step i should match input of step i+1)
            if i < len(arb_params.swap_steps) - 1:
                next_step = arb_params.swap_steps[i + 1]
                if step.token_out != next_step.token_in:
                    return False, f"Token mismatch between steps {i} and {i+1}"
        
        return True, "Validation passed"
    
    def execute_arbitrage(
        self,
        arb_params: ArbParams,
        gas_price: Optional[int] = None,
        dry_run: bool = False
    ) -> Dict[str, Any]:
        """
        Execute arbitrage transaction.
        
        Args:
            arb_params: Arbitrage parameters
            gas_price: Gas price in Wei (optional)
            dry_run: If True, simulate without executing
            
        Returns:
            Execution result dictionary
        """
        # Validate parameters
        is_valid, error_msg = self.validate_arb_params(arb_params)
        if not is_valid:
            print(f"Validation failed: {error_msg}")
            self.execution_stats["failed_executions"] += 1
            return {
                "success": False,
                "error": error_msg,
                "tx_hash": None
            }
        
        # Estimate gas
        gas_limit = self.estimate_gas(arb_params)
        
        # Encode swap steps
        encoded_steps = self.encode_swap_steps(arb_params.swap_steps)
        
        if dry_run:
            print(f"DRY RUN: Would execute arbitrage with {len(arb_params.swap_steps)} steps")
            print(f"Expected profit: {arb_params.expected_profit} Wei")
            return {
                "success": True,
                "dry_run": True,
                "gas_limit": gas_limit,
                "expected_profit": arb_params.expected_profit
            }
        
        # Execute transaction (placeholder for actual Web3 call)
        try:
            # In production, this would be:
            # tx = self.contract.functions.executeArbitrage(
            #     arb_params.flash_loan_amount,
            #     arb_params.flash_loan_token,
            #     arb_params.flash_loan_pool,
            #     encoded_steps,
            #     arb_params.deadline
            # ).transact({'gas': gas_limit, 'gasPrice': gas_price})
            
            tx_hash = "0x" + "0" * 64  # Placeholder
            
            print(f"Executing arbitrage transaction: {tx_hash}")
            print(f"Flash loan: {arb_params.flash_loan_amount} Wei of {arb_params.flash_loan_token}")
            print(f"Swap steps: {len(arb_params.swap_steps)}")
            print(f"Expected profit: {arb_params.expected_profit} Wei")
            
            # Update statistics
            self.execution_stats["total_executions"] += 1
            self.execution_stats["successful_executions"] += 1
            self.execution_stats["total_gas_used"] += gas_limit
            self.execution_stats["total_profit"] += arb_params.expected_profit
            
            return {
                "success": True,
                "tx_hash": tx_hash,
                "gas_limit": gas_limit,
                "expected_profit": arb_params.expected_profit,
                "swap_steps": len(arb_params.swap_steps)
            }
            
        except Exception as e:
            print(f"Execution failed: {str(e)}")
            self.execution_stats["failed_executions"] += 1
            return {
                "success": False,
                "error": str(e),
                "tx_hash": None
            }
    
    def get_execution_stats(self) -> Dict[str, Any]:
        """
        Get execution statistics.
        
        Returns:
            Dictionary of execution statistics
        """
        success_rate = (
            self.execution_stats["successful_executions"] / 
            max(self.execution_stats["total_executions"], 1) * 100
        )
        
        return {
            **self.execution_stats,
            "success_rate": success_rate,
            "avg_gas_per_execution": (
                self.execution_stats["total_gas_used"] / 
                max(self.execution_stats["total_executions"], 1)
            )
        }
